pkg/machine/cleanup.go

```diff
@@ -0,0 +1,66 @@
+package machine
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CleanupCallback struct {
+	Funcs []func() error
+	mu    sync.Mutex
+}
+
+func (c *CleanupCallback) Clean(err *error) {
+	// do not remove created files if the init is successful
+	if *err == nil {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+}
+
+func (c *CleanupCallback) CleanOnSignal(err *error) {
```

Comment:
With the suggestions described below I would recommend dropping the error argument (no longer needed)